core/src/main/scala-2.13+/cats/data/NonEmptyLazyList.scala

```diff
@@ -47,11 +47,13 @@ object NonEmptyLazyList extends NonEmptyLazyListInstances {
     s.asInstanceOf[LazyList[A]]
 
   def fromLazyList[A](as: LazyList[A]): Option[NonEmptyLazyList[A]] =
-    if (as.nonEmpty) Option(create(as)) else None
+    if (as.nonEmpty) Some(create(as)) else None
 
-  def fromLazyListUnsafe[A](ll: LazyList[A]): NonEmptyLazyList[A] =
-    if (ll.nonEmpty) create(ll)
-    else throw new IllegalArgumentException("Cannot create NonEmptyLazyList from empty LazyList")
+  def fromLazyListUnsafe[A](ll: LazyList[A]): NonEmptyLazyList[A] = {
+    @inline def ex = new IllegalArgumentException("Cannot create NonEmptyLazyList from empty LazyList")
+    if (ll.knownSize == 0) throw ex
+    else create(LazyList.empty #::: { if (ll.isEmpty) throw ex else ll })
+  }
```

Comment:
I see it was already discussed in this [thread](https://github.com/typelevel/cats/pull/4504#discussion_r1314308117),
with @armanbilge, but let me chime in a little bit too.

I'm thinking about a scenario like this one:
1. One creates a new NELL out of LL with the `fromLazyListUnsafe`
   ```scala
   def ll: LazyList[Int] = ???
   val nell = NonEmptyList.fromLazyListUnsafe(ll)
   ```
2. Now they pass it to a class that is implemented somewhere else (probably in some other library):
   ```scala
   class Foo[F[_]](nell: F[Int])(implicit F: Reducible[F]) {
     def calculateSomething: Int = nell.reduce
   }
   
   val foo = new Foo(nell)
   ```
3. The `Foo`'s method is called somewhere else as well:
    ```scala
    println(foo.calculateSomething)
    ```

The problem here is that `Foo` has no idea that it was fed with the "unsafe" NELL – it only knows that it gets `Reducible` which is guaranteed to be non-empty by design. Therefore, if to make such a thing possible, that means that we cannot rely on it anymore – i.e. any `Reducible` can throw upon any of its `reduce*` operations.

Reply:
> The problem here is that `Foo` has no idea that it was fed with the "unsafe" NELL

Sure, but this is fundamental for any lazy structure. E.g. someone can do something naughty inside of an `Eval` that causes it to throw and feed it to some innocent API that then gets an ugly surprise.

If a library developer uses "unsafe" wrong that's on them. We rely on "unsafe" layers all the time `IORuntime`, JVM, what-have-you to behave correctly. Users have to trust this stuff.

Reply:
Ok, makes sense, thanks.